### ISSUE TITLE
[Feature request] Project settings and custom telescope commands

### ISSUE THREAD

**Issue by ybbond** [akinsho/flutter-tools.nvim#79]
hi, is there config to add default args for FlutterRun? Also, is there a way to add custom menu to aforementioned FlutterRun (with default args) for Telescope?

I currently use these configuration to achieve what I wanted:

in the config for `flutter-tools.nvim`
```lua
on_attach = function(client, bufnr)
  -- .....
  require('flutter-tools.utils').command("FlutterAgentStart", [[lua require('flutter-tools.commands').run_command('--flavor=staging --no-sound-null-safety')]])
  -- .....
end
```

I copied the whole `flutter-tools`'s file `menu.lua`, then added these lines:
```lua
      {
        id = "flutter-tools-run-agent",
        label = "Flutter tools: Start Agent",
        hint = "Start the Agent App project",
        command = function() require("flutter-tools.commands").run_command('--flavor=staging --no-sound-null-safety') end,
      },
```

then added my own telescope folder in:
```
lua
|- configs
|- telescope
    |- _extensions
        |- flutter.lua
```

Is there "better" way to achieve this? If there is currently none, this is a feature request 😄 

Thanks before!

**Comment by akinsho** [comment]
Hi @ybbond,

Nope there currently isn't a way to do this, I discussed this a little bit over in #71. It isn't immediately clear how to do this easily tbh so I'm open to suggestions or PRs. I'm wary of adding in the wrong mechanism for this that would need to be deprecated later. For example vscode has a `launch.json` which a user can create per directory and so that informs what gets added to the run command etc. The problem with that is I'd have to create a custom per project config finding mechanism and avoid taken names like launch.json in case a user's project has one but they didn't intent to use it. Which feels like quite a bit of a scope creep into project settings territory for just one language specific plugin.

Alternatively it could be added to user's config but this then means it isn't project specific so you'd see run commands for other projects whilst in non-applicable projects. The solution could potentially be to namespace them under a particular directory in the config but then a user would have to add their project directory paths into their nvim config which is messy and maybe shares work information they don't want to share.

So TLDR, I'm not sure how to solve this problem.

**Comment by akinsho** [comment]
@Charly6596 do you happen to have any thoughts? I guess one thing could be to try and mirror the behaviour or reading values out of the `launch.json` although another completely hypothetical solution is I believe there's a stealth nvim plugin being worked on that will allow setting project based config for lsp more easily. It might be also worth waiting to see if that gets released since if there was a single plugin that all other plugins could depend on e.g. we had our own common format like `coc-settings.json` then I could reliably read that.

**Comment by Charly6596** [comment]
@akinsho hmm that nvim plugin sounds interesting +1 to waiting for it
I can't think about a clean solution either. Reading values from a launch.json file is fine I guess, but I prefer a more common/standard approach.
Using my own r command in my config worked for me so far, the same as what @ybbond did

**Comment by ybbond** [comment]
sorry just get back to this. I personally like the approach mentioned in https://github.com/akinsho/flutter-tools.nvim/issues/70#issuecomment-873615994. It makes the configuration has a single source of truth.

The problem that I can think with vscode's in-project-directory specific configuration is too scattered. Sublime's specific configuration that doesn't need (but can) to be inside the project directory is better, but still too scattered for me.

**Comment by akinsho** [comment]
So I'm going to wait a little for the possible project/workspace plugin that might be being developed to get released. I don't actually like the idea of putting project specific configuration in your dotfiles, I think this leaks potentially private information regarding a user's projects or means adding not globally relevant information to a user's dotfiles. I think the config in directory thing is a well known mechanism, and I'd rather use something like that.

**Comment by ybbond** [comment]
@akinsho that makes sense, I respect your decision.

I guess it leaves us to the "custom command" part? You're welcome to focus this issue on the project specific part

**Comment by akinsho** [comment]
@ybbond I think those would be intertwined right, since the need for a custom command relates to a project specific requirement. It needs a bit of thought in general, since I'm not sure what the best interface will be. I would rather users didn't end up specifying like 7 different variants of flutter run for a bunch of different projects, whereas they could configure a project and its run args or something. A naive solution would be to specify some list of custom commands in a user's config with the necessary metadata and then just merge those with the existing list when telescope is opened, but I'd rather get it right and not be iterating or dealing with bugs or requests for ages.

**Comment by akinsho** [comment]
Not that I've got mountain of times to add it in right now but jotting down for reference. Now that `exrc` is supported again users can create a `.nvim.lua` or `.nvimrc` which means that all this would now need is some mechanism for adding project config e.g. there could be an additional API function like
```lua
require('flutter-tools').setup_project({
  device = device_choice,
  flavors = {},
})
```

essentially there could be some object containing the metadata for this feature passed to that function that is then used to configure the project.

**Comment by sidlatau** [comment]
Oh, nice! I want to have the option to choose which flavor to use.
Example config:
```lua
require('flutter-tools').setup_project({
  run_configuration = {
    {
       name="Flavor 1",
       args= { "--dart-define=value=value1 --flavor value1" }
    },
     {
       name="Flavor 2",
       args= { "--dart-define=value=value2 --flavor value2" }
    },
  },
})
```
If a such configuration exists and there are 2 or more configurations, then when running `FlutterRun` it may show a select popup (similar to that used for devices/emulators) to choose the desired configuration to run.
It will then run:
```
FlutterRun --dart-define=value=value1 --flavor value1
```
If there is only one configuration, then it will automatically add arguments to `FlutterRun` command.
What do you think @akinsho about extending `FlutterRun` in such way?

**Comment by akinsho** [comment]
@sidlatau that sounds good in terms of configurability, I started on a basic version locally but didn't account for multiple configurations in this way but can easily extend it. I was thinking more of using structured data though with the options that the key maps to the value and it's handled but the plugin e.g. 

```lua
require('flutter-tools').setup_project({
	{
		device = "chrome",
		flavor = "flavor1",
		dart_define = {
			key1 = value
		}
	},
	{
		device = "pixel6",
		flavor = "flavor2",
		dart_define = {
			key2 = value2
		}
	}
})
```

This way we can easily programatically do stuff based on the values a user sets so e.g. we could validate the `device` or maybe show the flavor is a statusline component. Basically I'd rather avoid lists of arbitrary strings that we:
 a) can't inspect or use
 b) can't validate easily
 c) would have to parse if we wanted to do a or b
It also means for a dev perspective the object can be properly type annotated and would be nicer to work with

**Comment by akinsho** [comment]
Going to close this out now since #232 go merged, there's always scope for expanding what exists but the main issue of project specific settings I think at least has a path forward

**Comment by nank1ro** [comment]
@akinsho It would be very useful to be able to insert args in the configuration, specifically the target argument is missing and it would be great if it could be inserted.
What do you think? In case I try to open a PR with the modification.

**Comment by akinsho** [comment]
@nank1ro by all means please raise a PR generally the settings are intended to be expanded when use cases arise so feel free to open one for the target argument
